[LV] Pre-committing tests for changing loop interleaving count computation

nilanjana87 opened the pull request:
Added tests for evaluating changes to loop interleaving count computation and for removing loop interleaving threshold.

nilanjana87 committed:
[LV] Pre-committing tests for changing interleaving count computation

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/interleave_count.ll:
```diff
@@ -0,0 +1,106 @@
+; RUN: opt < %s -force-vector-width=64 -O3 -S -pass-remarks=loop-vectorize 2>&1 | FileCheck %s
```
please don't use `-O3`, as this makes the test prone to failing due to changes in other parts of the pipeline.

Also, what's the motivation for forcing vector-width to 64 here? It is an impossible VF for AArch64. Would it be possible to just let the cost model also pick the VF?

nilanjana87 replied:
> Also, what's the motivation for forcing vector-width to 64 here? It is an impossible VF for AArch64. Would it be possible to just let the cost model also pick the VF?

The purpose of the test is to show the impact of interleaving count computation change that will be added in a subsequent patch https://github.com/llvm/llvm-project/pull/70141. The IC is a function of VF & TC (trip count) and to test this change the VF needs to be large enough that the vectorized-interleaved loop runs once or twice. However, the TC threshold to allow interleaving is 128, so I increased the VF to hit this case. 

On trying to work around this by reducing the TC & the TC threshold with `-tiny-trip-count-interleave-threshold` option, the loop is not getting vectorized because of hitting a clause of runtime pointer checks.

fhahn replied:
Ah I see. The runtime check issue should be resolvable by `define void @loop_with_tc_128(ptr noalias %p, ptr noalias %q) {`

nilanjana87 replied:
I reduced the VF to 8 & made relevant changes to show the same intended impact. 
I forced a VF because otherwise the vectorization threshold will come into play, adding one more configuration option. For a VF smaller than 8, I'll have to tune other parameters like small-loop-cost to make these tests show the IC computation without getting blocked earlier.

fhahn replied:
IIUC this test should ideally check the cost model selecting the expected VF and IC. Forcing a VF seems to leave a gap in testing here and it would be better to not artificially force the VF, in particular because the cos model is expected to pick 16 for the loops?

nilanjana87 replied:
Increased the loop trip count to hit the intended case of changed interleave count computation in subsequent patches without forcing VF.

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/interleave_count.ll:
```diff
@@ -0,0 +1,106 @@
+  ret void
+}
+
+; For a loop with unknown trip count but a profile showing an approx TC estimate of 129, 
```
This is still a todo IIUC, please add `TODO` to avoid confusion

nilanjana87 replied:
Done

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/interleave_count.ll:
```diff
@@ -0,0 +1,106 @@
+  ret void
+}
+
+; For a loop with known trip count of 129, when we force VF 64, it should use
```
needs `TODO:`

nilanjana87 replied:
Done

fhahn commented on llvm/test/Transforms/LoopVectorize/X86/unroll-small-loops.ll:
```diff
@@ -32,7 +32,193 @@ define i32 @foo(ptr nocapture %A) nounwind uwtable ssp {
+  br i1 %exitcond, label %5, label %1
+
+; <label>:5                                       ; preds = %1
+  ret i32 undef
```
Better use `ret void`

nilanjana87 replied:
Done

fhahn commented on llvm/test/Transforms/LoopVectorize/X86/unroll-small-loops.ll:
```diff
@@ -32,7 +32,193 @@ define i32 @foo(ptr nocapture %A) nounwind uwtable ssp {
+define i32 @foo_trip_count_16(ptr nocapture %A) nounwind uwtable ssp {
+  br label %1
+
+; <label>:1                                       ; preds = %1, %0
```
please use named labels for the blocks here and for the exit block

nilanjana87 replied:
Done. I did this manually using string replacement for renaming the blocks & associated re-numbering for the variables, but I'm wondering if there's a tool/option to do this automatically?

fhahn reviewed (commented):
Thanks for the additional test coverage. Some comments inline

nilanjana87 committed:
Addressed reviewer comments

- replaced -O3 in tests with -p loop-vectorize
- added "TODO" in comments for tests that are meant to change after the subsequent patch
- used named labels for basic blocks

fhahn commented on llvm/test/Transforms/LoopVectorize/X86/unroll-small-loops.ll:
```diff
@@ -6,37 +6,257 @@
+  ret void
+}
+
+; TODO: We unroll this loop 4 times since TC not being a multiple of VF may require
```
Should this be `TODO: We *should* unroll...`? As at the moment it's not unrolled at all right?

nilanjana87 replied:
yes. Fixed it.

nilanjana87 committed:
Changed test VFs to a more reasonable value, fixed comments

nilanjana87 committed:
Changed trip count for the added test cases to test the IC computation change in subsequent patches without the need to force VF

fhahn reviewed (approved):
LGTM, thanks!

nilanjana87 committed:
Disabled redundant output in tests

fhahn reviewed (approved):
LGTM, thanks!
